### DISCUSSION TITLE
Get the required fields of a complex schema (variants and unions)

### DISCUSSION THREAD

**Discussion by ruiaraujo012** [open-circle/valibot#484] [Q&A]
Hello, I'm building a React form with some complex validations, like if `foo` is true, `bar` is required, otherwise is optional.
To do that, I need to use unions and variants, and here is the problem, how can extract all the fields that are required, so I can mark the form field as required (*)? Any help please?

Thank you.

**Comment by fabian-hiller** [comment]
Object entries are required by default. I recommend to detect if a field is optional and then remove the (*).

**Reply by ruiaraujo012** [reply]
Ok, yes, it could do, but how can I know that the schema is optional? What I'm doing is, each input (custom component) receive the schema and his path, and then, the (*) appears or not based on the schema, the problem is that variants and unions are a bit complicated to parse than regular objects.

**Reply by ruiaraujo012** [reply]
One example is:
```js
const IsMilitarySchema = v.object({
  isMilitary: v.literal(true),
  militaryNumber: v.string([v.minLength(1)])
});

const NotMilitarySchema = v.object({
  isMilitary: v.literal(false),
  militaryNumber: v.optional(v.string())
});

const MilitarySchema = v.variant('isMilitary', [IsMilitarySchema, NotMilitarySchema])

const PTPersonSchema = v.object({
  country: v.literal('PT'),
  address: v.string([v.minLength(1)]) 
  })

const NotPTPersonSchema = v.object({
  country: v.string([v.minLength(1)]),
  address: v.optional(v.string())
  })

const PersonSchema = v.union([PTPersonSchema, NotPTPersonSchema])

const Schema = v.object({
    military: MilitarySchema,
    person: PersonSchema
  }) 
```

**Reply by fabian-hiller** [reply]
You need to write a recursive function similar to the code of `getDefaults`: https://github.com/fabian-hiller/valibot/blob/v0.30.0/library/src/methods/getDefaults/getDefaults.ts

**Reply by ruiaraujo012** [reply]
Ok, and if it is of type object, I can find if it is optional, otherwise, is required. It could do. If I rerun that function when some conditions happen, will it select the correct element in union/variant?

**Reply by fabian-hiller** [reply]
For any complex schema you need to add its type and write a custom implementation. For `object` you need to handle its `.entries` and for `union` and `variant` its `.options`. If you can't figure it out I can try to make time on the weekend to provide you sample code.

**Reply by ruiaraujo012** [reply]
Thank you for your help, I'm going to try making a function to return the optional fields and if I went in some problem or something I'll ask for your advice 😅 .

**Reply by ruiaraujo012** [reply]
I managed to put it to work (partially), and it is here where I'm confused, and the things got complicated. With the code below, I get this:
```js
{
    "military": {
        "0": {},
        "1": {
            "militaryNumber": true
        }
    },
    "person": {
        "0": {},
        "1": {
            "address": true
        }
    }
}
```

How can I know what is the schema (union or variant) applied to know if the field is or not optional?

Code:

```ts
import * as v from 'valibot';

const IsMilitarySchema = v.object({
  isMilitary: v.literal(true),
  militaryNumber: v.string([v.minLength(1)]),
});

const NotMilitarySchema = v.object({
  isMilitary: v.literal(false),
  militaryNumber: v.optional(v.string()),
});

const MilitarySchema = v.variant('isMilitary', [IsMilitarySchema, NotMilitarySchema]);

const PTPersonSchema = v.object({
  address: v.string([v.minLength(1)]),
  country: v.literal('PT'),
});

const NotPTPersonSchema = v.object({
  address: v.optional(v.string()),
  country: v.string([v.minLength(1)]),
});

const PersonSchema = v.union([PTPersonSchema, NotPTPersonSchema]);

const Schema = v.object({
  military: MilitarySchema,
  person: PersonSchema,
});

type PossibleSchemaType =
  | v.BaseSchema
  | v.ObjectSchema<v.ObjectEntries, any>
  | v.VariantSchema<any, any>
  | v.UnionSchema<any, any>;

function getOptionals<TSchema extends PossibleSchemaType>(schema: TSchema): Record<string, unknown> | undefined {
  if (v.isOfType('object', schema)) {
    return Object.fromEntries(
      Object.entries(schema.entries).map(([key, value]) => [
        key,
        v.isOfType('optional', value) ? true : getOptionals(value),
      ]),
    );
  } else if (v.isOfType('union', schema) || v.isOfType('variant', schema)) {
    return Object.fromEntries(
      Object.entries(schema.options).map(([key, value]) => [key, getOptionals(value as PossibleSchemaType)]),
    );
  }

  return undefined;
}

const optionalFields = getOptionals(Schema);

console.log('optionalFields >> ', optionalFields);
```

**Reply by fabian-hiller** [reply]
> How can I know what is the schema (union or variant) applied to know if the field is or not optional?

I think this is not possible because it is decided dynamically at runtime. I would try to store some information (e.g. the path to the field) in an array or object that allows you to identify each optional field. At the time you render your form, you can use this information to know if a field is optional.

```ts
// Using an array
const optionalFields: string[] = ['military.1.militaryNumber', 'person.1.address'];
const isOptional = (path: string) => optionalFields.includes(path);

// Using an object
const optionalFields: Record<string, boolean> = {
  'military.1.militaryNumber': true,
  'person.1.address': true,
};
const isOptional = (path: string) => !!optionalFields[path];
```

**Reply by ruiaraujo012** [reply]
I think you're right, there's no way to do that.

Like the angular forms, is there any options to dynamically add/remove validations with valibot?

**Reply by fabian-hiller** [reply]
Not sure but `special` or `lazy` could work but I only recommend to use them for very special cases or for recursive schemas.

- https://valibot.dev/guides/other/
- https://valibot.dev/api/special/
- https://valibot.dev/api/lazy/

**Reply by ruiaraujo012** [reply]
Well, I gave up from the idea of trying to extract the optional fields from the schema and instead, I pass the optional prop conditionally based on the data to the component that I want. Thanks for your help, it helped me to reach this approach :)

**Reply by fabian-hiller** [reply]
I do the same. As long as your form isn't crazy complex, this is fine.
